File: py_validation/pardox/_io/rest.py
import json
from ..wrapper import lib


def read_rest(url: str, headers: dict = None):
    """
    Fetch JSON data from a REST API endpoint into a DataFrame.
    The endpoint must return a JSON array of objects.

    Example:
        df = pdx.read_rest("https://api.example.com/sales")
        print(df)
    """
    from ..frame import DataFrame
    if not hasattr(lib, 'pardox_read_rest'):
        raise NotImplementedError("pardox_read_rest not found in Core.")
    headers_json = json.dumps(headers or {}).encode('utf-8')
    ptr = lib.pardox_read_rest(url.encode('utf-8'), headers_json)
    if not ptr:
        raise RuntimeError(f"read_rest failed for URL: {url}")
    return DataFrame(ptr)

File: js_validation/pardox/src/_io/rest.js
'use strict';

const { getLib } = require('../ffi');
const { DataFrame } = require('../DataFrame');

/**
 * Fetch JSON data from a REST API endpoint into a DataFrame.
 * @param {string} url      REST endpoint URL.
 * @param {string} method   HTTP method (default 'GET').
 * @param {Object} headers  Optional HTTP headers.
 * @returns {DataFrame}
 */
function readRest(url, method = 'GET', headers = {}) {
    const lib = getLib();
    const ptr = lib.pardox_read_rest(url, method, JSON.stringify(headers));
    if (!ptr) throw new Error(`readRest failed for URL: ${url}`);
    return new DataFrame(ptr);
}

module.exports = { readRest };
